internal/provider/atlas_schema_resource.go

```diff
@@ -0,0 +1,347 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"ariga.io/atlas/sql/schema"
+	"ariga.io/atlas/sql/sqlclient"
+	"github.com/hashicorp/hcl/v2/hclparse"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type (
+	// AtlasSchemaResource defines the resource implementation.
+	AtlasSchemaResource struct{}
+	// AtlasSchemaResourceModel describes the resource data model.
+	AtlasSchemaResourceModel struct {
+		ID     types.String `tfsdk:"id"`
+		HCL    types.String `tfsdk:"hcl"`
+		URL    types.String `tfsdk:"url"`
+		DevURL types.String `tfsdk:"dev_db_url"`
+	}
+)
+
+// Ensure provider defined types fully satisfy framework interfaces
+var (
+	_ resource.Resource                   = &AtlasSchemaResource{}
+	_ resource.ResourceWithModifyPlan     = &AtlasSchemaResource{}
+	_ resource.ResourceWithValidateConfig = &AtlasSchemaResource{}
+)
+
+func NewAtlasSchemaResource() resource.Resource {
+	return &AtlasSchemaResource{}
+}
+
+func (r *AtlasSchemaResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_schema"
+}
+
+func (r *AtlasSchemaResource) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diagnostics) {
+	return tfsdk.Schema{
+		Description: "Atlas database resource manages the data schema of the database, " +
+			"using an HCL file describing the wanted state of the database. " +
+			"See https://atlasgo.io/",
+		Attributes: map[string]tfsdk.Attribute{
+			"hcl": {
+				Description: "The schema definition for the database " +
+					"(preferably normalized - see `atlas_schema` data source)",
+				Type:     types.StringType,
+				Required: true,
+			},
+			"url": {
+				Description: "The url of the database see https://atlasgo.io/cli/url",
+				Type:        types.StringType,
+				Required:    true,
+				Sensitive:   true,
+			},
+			"dev_db_url": {
+				Description: "The url of the dev-db see https://atlasgo.io/cli/url",
+				Type:        types.StringType,
+				Optional:    true,
+				Sensitive:   true,
+			},
+			"id": {
+				Description: "The ID of this resource",
+				Computed:    true,
+				PlanModifiers: tfsdk.AttributePlanModifiers{
+					resource.UseStateForUnknown(),
+				},
+				Type: types.StringType,
+			},
+		},
+	}, nil
+}
+
+func (r *AtlasSchemaResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data *AtlasSchemaResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
```

Comment:
@giautm i see this repeat in many places, isn't it nicer to do:
```suggestion
	// Read Terraform plan data into the model
	if diags :=  req.Plan.Get(ctx, &data); diags.HasErrors() {
	  resp.Diagnostics.Append(diags...)
	  return
	}
```

Reply:
This code was generated as example from terraform.

Reply:
@rotemtam: I think the original code is more correct because diagnostics may contain warnings and should be added to the result, we only stop progress on the errors.